File: src/PublisherEntity.java
import Helpers.BusReader;
import Models.Broker;
import Models.Publisher;
import Models.Stigma;
import Models.Wrapper;

import java.io.ObjectOutputStream;
import java.net.InetAddress;
import java.net.Socket;
import java.util.concurrent.TimeUnit;

class PublisherEntity {

    private Publisher publisher;

    PublisherEntity(Publisher publisher) {
        this.publisher = publisher;
    }

    void start() {
        // Start the listener, to listen from brokers
        new PublisherHandler(this.publisher).start();

        try {
            for (int busID : this.publisher.getTopics()) {
                for (Stigma stigma : BusReader.readBusPositions(busID)) {
                    TimeUnit.SECONDS.sleep(3);
                    push(busID, stigma);
                    System.out.println(this.publisher.toString() + " Got data from sensor -> " + busID + ": " + stigma);
                }
            }
        } catch (Exception err) {
            err.printStackTrace();
        }
    }

    private void push(int topic, Stigma value) {
        for (Broker broker : this.publisher.getBrokers()) {
            System.out.println("Checking " + broker.toString());
            if (publisher.hashThis(String.valueOf(topic)).compareTo(broker.getHash()) > 0) {
                sendToBroker(broker, value);
            }
        }
    }

    private void sendToBroker(Broker broker, Stigma stigma){
        try {
            Socket requestSocket;
            ObjectOutputStream out;
            requestSocket = new Socket(InetAddress.getByName(broker.getIP()), broker.getPort());
            out = new ObjectOutputStream(requestSocket.getOutputStream());
            Wrapper<Stigma> toSend = new Wrapper<>();
            toSend.data = stigma;
            System.out.println(this.publisher.toString() + " Sent -> " + toSend);
            out.writeUnshared(toSend);
            out.flush();
        } catch (Exception err) {
            System.err.println(this.publisher.toString() + " Tried to connect to -> " + broker.toString() + " But was down.");
        }
    }

    PublisherEntity addTopic(int busLineID) {
        this.publisher.addTopic(busLineID);
        return this;
    }

}

File: src/entities/PublisherEntity.kt
package entities

import handlers.PublisherHandler
import helpers.BusProvider
import models.Broker
import models.Publisher
import models.Stigma
import models.Wrapper

import java.io.ObjectOutputStream
import java.net.InetAddress
import java.net.Socket
import java.util.concurrent.TimeUnit

class PublisherEntity(private val publisher: Publisher) {

    fun start() {

        // Start the listener, to listen from brokers
        PublisherHandler(publisher).start()

        // Supply data once
        publisher.data = BusProvider.readBusPositions(publisher.topics.toIntArray())

        // Distribute data for each Bus synchronously
        while (publisher.data.isNotEmpty()) {
            for (key in publisher.data.keys) {
                TimeUnit.SECONDS.sleep(2)
                val stigma: Stigma = publisher.data[key]!!.poll()
                stigma.topic = key
                if (!push(key, stigma)) {
                    println("$publisher Got data from sensor -> $key: $stigma")
                }
            }
        }

        // Bus got lost in a black hole
        println("$publisher ran out of data!")

    }

    private fun push(topic: Int, value: Stigma): Boolean {
        for (broker in this.publisher.brokers) {
            if ((topic.toString()).hashCode() >= broker.hash) {
                sendToBroker(broker, value)
                return true
            }
        }
        return false
    }

    private fun sendToBroker(broker: Broker, stigma: Stigma) {
        try {
            val requestSocket = Socket(InetAddress.getByName(broker.ip), broker.port)
            val out: ObjectOutputStream
            out = ObjectOutputStream(requestSocket.getOutputStream())
            val toSend = Wrapper<Stigma>()
            toSend.data = stigma
            println("$publisher Sent -> $toSend to $broker")
            out.writeUnshared(toSend)
            out.flush()
        } catch (err: Exception) {
            System.err.println(this.publisher.toString() + " Tried to connect to -> " + broker.toString() + " But was down.")
        }

    }

    fun addTopic(busLineID: Int): PublisherEntity {
        this.publisher.addTopic(busLineID)
        return this
    }

}
